sdk/ai/azure-ai-inference/src/main/java/com/azure/ai/inference/ChatCompletionClientTracer.java

```diff
@@ -0,0 +1,606 @@
+                final RequestOptions rOptions = resource.requestOptions.setContext(resource.span);
+                final Flux<StreamingChatCompletionsUpdate> completionChunks
+                    = resource.operation.invoke(resource.completeRequest, rOptions);
+                return completionChunks.doOnNext(resource::onNextChunk);
+            };
+
+        final Function<StreamingChatCompletionsState, Mono<Void>> asyncComplete = resource -> {
+            final Context span = resource.span;
+            final StreamingChatCompletionsUpdate lastChunk = resource.lastChunk;
+            final String finishReasons = resource.getFinishReasons();
+
+            traceCompletionResponseAttributes(lastChunk, finishReasons, span);
+            traceChoiceEvent(resource.toJson(logger), OffsetDateTime.now(ZoneOffset.UTC), span);
+            tracer.end(null, null, span);
+            return Mono.empty();
+        };
+
+        final BiFunction<StreamingChatCompletionsState, Throwable, Mono<Void>> asyncError = (resource, throwable) -> {
+            final Context span = resource.span;
+
+            tracer.setAttribute("error.type", throwable.getClass().getName(), span);
+            traceChoiceEvent(FINISH_REASON_ERROR, OffsetDateTime.now(ZoneOffset.UTC), span);
+            tracer.end(null, throwable, span);
+            return Mono.empty();
+        };
+
+        final Function<StreamingChatCompletionsState, Mono<Void>> asyncCancel = resource -> {
+            final Context span = resource.span;
+            tracer.setAttribute("error.type", "cancelled", span);
+            traceChoiceEvent(FINISH_REASON_CANCELED, OffsetDateTime.now(ZoneOffset.UTC), span);
+            tracer.end("cancelled", null, span);
+            return Mono.empty();
+        };
+
+        return Flux.usingWhen(resourceSupplier, resourceClosure, asyncComplete, asyncError, asyncCancel);
+    }
+
+    //<editor-fold desc="Private util types, methods">
+    private String rootSpanName(ChatCompletionsOptions completeRequest) {
+        return CoreUtils.isNullOrEmpty(completeRequest.getModel()) ? "chat" : "chat " + completeRequest.getModel();
+    }
+
+    private void traceCompletionRequestAttributes(ChatCompletionsOptions request, Context span) {
+        final String modelId = request.getModel();
+        tracer.setAttribute("gen_ai.operation.name", "chat", span);
+        tracer.setAttribute("gen_ai.system", INFERENCE_GEN_AI_SYSTEM_NAME, span);
+        tracer.setAttribute("gen_ai.request.model", CoreUtils.isNullOrEmpty(modelId) ? "chat" : modelId, span);
+        if (request.getMaxTokens() != null) {
+            tracer.setAttribute("gen_ai.request.max_tokens", request.getMaxTokens(), span);
+        }
+        if (request.getTemperature() != null) {
+            tracer.setAttribute("gen_ai.request.temperature", request.getTemperature(), span);
+        }
+        if (request.getTopP() != null) {
+            tracer.setAttribute("gen_ai.request.top_p", request.getTopP(), span);
```

Comment:
There are a few more attributes in the latest semantic conventions - https://github.com/open-telemetry/semantic-conventions/blob/v1.29.0/docs/gen-ai/azure-ai-inference.md. It's ok to deviate from python impl (there are no breaking changes in the attributes)

Reply:
covered other attributes as well.

Reply:
thank you!